File: src/main/java/cookiedragon/luchadora/mixin/mixins/lwjgl/MixinDisplay.java
package cookiedragon.luchadora.mixin.mixins.lwjgl;

import cookiedragon.luchadora.event.api.EventDispatcher;
import cookiedragon.luchadora.event.lwjgl.SetDisplayTitleEvent;
import org.lwjgl.opengl.Display;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

/**
 * @author cookiedragon234 12/Jan/2020
 */
@Mixin(Display.class)
public class MixinDisplay
{
	@Shadow(remap = false)
	private static String title;
	
	@Inject(method = "setTitle", at = @At("HEAD"), cancellable = true, remap = false, require = 1)
	private static void onSetTitle(String newTitle, CallbackInfo ci)
	{
		System.out.println("Set title " + newTitle + " : " + title);
		SetDisplayTitleEvent event = new SetDisplayTitleEvent(title, newTitle);
		EventDispatcher.dispatch(event);
		if (!event.getNewTitle().equals(newTitle))
		{
			ci.cancel();
			Display.setTitle(event.getNewTitle());
		}
	}
}

File: src/main/kotlin/cookiedragon/luchadora/mixin/mixins/lwjgl/MixinDisplay.kt
package cookiedragon.luchadora.mixin.mixins.lwjgl

import cookiedragon.eventsystem.EventDispatcher
import cookiedragon.luchadora.event.lwjgl.SetDisplayTitleEvent
import org.lwjgl.opengl.Display
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

/**
 * @author cookiedragon234 15/Feb/2020
 */
@Mixin(Display::class)
class MixinDisplay {
	@Shadow(remap = false)
	private val title: String? = null
	
	@Inject(method = ["setTitle"], at = [At("HEAD")], cancellable = true, remap = false, require = 1)
	private fun onSetTitle(newTitle: String, ci: CallbackInfo) {
		with(SetDisplayTitleEvent(title ?: "", newTitle)) {
			EventDispatcher.dispatch(this)
			if (newTitle != newTitle) {
				ci.cancel()
				Display.setTitle(newTitle)
			}
		}
	}
}
